src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,731 @@
+          RestrictedPathGlob::Wildcard { wildcard } => wildcard,
+          RestrictedPathGlob::DirWildcard { wildcard, .. } => wildcard,
+        };
+
+        let matching_files: Vec<PathBuf> = cur_dir_files
+          .keys()
+          .filter(|path| {
+            // NB: match just the current path component against the wildcard, but use the prefix
+            // when checking against the `exclude` patterns!
+            wildcard.matches_path(path) && !exclude.is_ignored_path(&prefix.join(path), false)
+          })
+          .cloned()
+          .collect();
+        for file_path in matching_files.into_iter() {
+          // NB: remove any matched files from `cur_dir_files`, so they are successfully matched
+          // against at most once.
+          let file_node = cur_dir_files.remove(&file_path).unwrap();
+          globbed_files.insert(file_path, file_node);
+        }
+
+        let matching_directories: Vec<PathBuf> = cur_dir_directories
+          .keys()
+          .filter(|path| {
+            // NB: match just the current path component against the wildcard, but use the prefix
+            // when checking against the `exclude` patterns!
+            wildcard.matches_path(path) && !exclude.is_ignored_path(&prefix.join(path), true)
+          })
+          .cloned()
+          .collect();
+        for directory_path in matching_directories.into_iter() {
+          // NB: do *not* remove the directory for `cur_dir_directories` after it is matched
+          // successfully once!
+          let directory_node = cur_dir_directories.get(&directory_path).unwrap();
+
+          // TODO(#9967): Figure out how to consume the existing glob matching logic that works on
+          // `VFS` instances!
+          match &path_glob {
+            RestrictedPathGlob::Wildcard { wildcard } => {
+              assert_ne!(*wildcard, *DOUBLE_STAR_GLOB);
+              // This directory matched completely, without having to subset its contents
+              // whatsoever. We can avoid traversing its contents and return the node.
+              // NB: This line should be idempotent if the same directory node is added twice.
+              globbed_directories.insert(directory_path, directory_node.clone());
+            }
+            RestrictedPathGlob::DirWildcard {
+              wildcard,
+              remainder,
+            } => {
+              let mut subdir_globs: Vec<RestrictedPathGlob> = vec![];
+              if *wildcard == *DOUBLE_STAR_GLOB {
+                // NB: here we short-circuit several cases which would swallow up a directory
+                // without subsetting it.
+                match &remainder[..] {
+                  // ending in /**
+                  [] => {
+                    globbed_directories.insert(directory_path, directory_node.clone());
+                    continue;
+                  }
+                  // ending in /**/* or /**/**
+                  [single_glob]
```

Comment:
I think that this might be looking ahead unnecessarily. This arm of the match represents looking both at the current and the next directory in the hierarchy to see if they are both wildcards. I think that if you were to change this `match &remainder[..]` into just `if remainder.is_empty`, you would apply this match on the next iteration of the loop...? A bit simpler.

Reply:
Done!